### ISSUE TITLE
model specs appear to retain values from data matrix

### ISSUE THREAD

**Issue by mmacpherson** [matthewwardrop/formulaic#33]
I was looking at the modelspec object after a `formulaic.model_matrix(formula, df)` call with some synthetic data, and noted that modelspec was rather large (self-contained code eg below). Poking at the object (using `pympler`), I found that the `EvaluatedFactor` terms nested within that `.structure` attribute contain copies of the data columns from the data matrix.

When I iterated through `.structure` and set the `EvaluatedFactor.values` terms to `None`, `.get_model_matrix()` continued to function correctly on a dataframe different from the original one I'd used to create the modelspec.

A few questions, if I might?

- Are there circumstances where it's necessary to retain those data copies in the modelspec object, in order to support subsequent `.get_model_matrix` calls?
- If not, might it be reasonable not to store those potentially-large data under the modelspec, or perhaps store them optionally, eg via some kwarg to the `model_matrix` method?
- Would there be anywhere else within the modelspec that data are stored? (From this experiment at least, it doesn't appear to be so.)

Here's the code I used, for reference:

```python
import numpy as np
import pandas as pd
import formulaic as fm
import pympler
import scipy.special

np.random.seed(123)

# -- a formula to exercise formulaic
formula = "y ~ x1 * x2 + bs(x1, 3) + m * x1"

n = 100000  # largish to make detection easy

def gen_frame(n):
    return pd.DataFrame(
        dict(
            x1=np.random.normal(size=n),
            x2=np.random.normal(size=n),
            m=np.random.choice(list("abc"), size=n),
            e=np.random.normal(size=n),
        )).assign(y=lambda f: (scipy.special.expit(1 + 2 * f.x1 - f.x2 + f.e) >
                               0.5).astype(int))

df = gen_frame(n)  # ie, training set
dfp = gen_frame(5)  # ie, prediction set

y, X = fm.model_matrix(formula, df)

f"Original ModelSpec size: {pympler.asizeof.asizeof(X.model_spec)}"

# -- build design matrix from original modelspec
Xp1 = X.model_spec.get_model_matrix(dfp)

# -- remove values from evaluated factors
for _, factors, _ in X.model_spec.structure:
    for f in factors:
        for ff in f.factors:
            ff.factor.values = None

f"Decanted ModelSpec size: {pympler.asizeof.asizeof(X.model_spec)}"

# -- build design matrix from decanted modelspec
Xp2 = X.model_spec.get_model_matrix(dfp)

# -- prove that the .values are unnecessary, at least in this case
pd.testing.assert_frame_equal(Xp1, Xp2)
```

which yields:

```
'Original ModelSpec size: 28436952'
'Decanted ModelSpec size: 17984'
```

and does not raise an assertion error, implying that the two design matrices had identical values.

Thank you for creating and sharing this package!

**Comment by matthewwardrop** [comment]
This is a bug! Thanks for pointing it out. I'll patch this data leak soon!

**Comment by matthewwardrop** [comment]
Thanks again for reporting this! I've patched this in 4141095493a56f6d340ca63c8c0266a81a216403. I'll put out Formulaic 0.2.4 imminently.

**Comment by mmacpherson** [comment]
Belatedly... Thank you very much for addressing this!

I confirm that the new version of formulaic resolves the issue demonstrated above.
